### PR TITLE
[next-server v8.1] Add useRouter() hook

### LINKED ISSUE #35019: [@types/next-server/router.d.ts] add `useRouter` definition
- [ ] Add useRouter hook definition (https://github.com/zeit/next.js/pull/6869)

- [ ] Overall follow the version `8.1.0` (currently it's `8.0.4`) ?

### PR DESCRIPTION
**Pull request by kachkaev**
Closes #35019 

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/zeit/next.js/pull/6869
- [x] Increase the version number in the header if appropriate.
- ~~If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.~~

cc @brikou, @jthegedus, @resir014, @scottdj92, @joaovieira, @ajliv

### CODE REVIEW

**Comment by JerryGreen** [comment]
Cool. Not sure if it's enough for bumping to 8.1, though...

**Comment by kachkaev** [comment]
The hook was added in Next 8.1, that's why I bumped: https://github.com/zeit/next.js/releases/tag/v8.1.0

If anything else is missing, we can release it as `@types/next-server@8.1.x`.

**Comment by JerryGreen** [comment]
@kachkaev `RouterProps` was correct. I don't have `readyCallbacks` or `ready` callback there:

```
asPath: "/login"
back: ƒ ()
beforePopState: ƒ ()
events: {on: ƒ, off: ƒ, emit: ƒ}
pathname: "/login"
prefetch: ƒ ()
push: ƒ ()
query: {}
reload: ƒ ()
replace: ƒ ()
route: "/login"
components: (...)
get components: ƒ ()
__proto__: Object
```

**Comment by kachkaev** [comment]
If someone can't wait for the changes to be merged and released, here is a temporary workaround that helped me for now. Simply create a new local file called `thirdParty.d.ts` with the following content:

```ts
// https://github.com/DefinitelyTyped/DefinitelyTyped/pull/35058

import { RouterProps } from "next-server";

declare module "next/router" {
  export function useRouter(): RouterProps;
}
```

**Comment by JerryGreen** [comment]
A note: There's a server part which is rendering the components too / executes the same code, and it receives a different (simpler) router format than browser:

```
ServerRouter {
  route: '/login',
  pathname: '/login',
  query: {},
  asPath: '/login' }
```

May be important too

File: types/next-server/router.d.ts [hunk 1]
```diff
@@ -96,5 +96,7 @@ export function withRouter<T extends {}, Q = DefaultQuery>(
   Component: React.ComponentType<T & WithRouterProps<Q>>
 ): React.ComponentType<Omit<T, keyof WithRouterProps<Q>>>;
 
+export function useRouter(): RouterProps;
```
Comment (resir014) [inline types/next-server/router.d.ts]:
```suggestion
export function useRouter<Q = DefaultQuery>(): RouterProps<Q>;
```

I think we should still include a generic here for passing the router query.

**Review by resir014 — CHANGES_REQUESTED** [review]
One tiny thing before this is good to go:

**Review by resir014 — APPROVED** [review]
Looks good now. Thanks @kachkaev!

**Review by scottdj92 — APPROVED** [review]

**Pull request merged**